### PR TITLE
CIV-10054 More Information Required HWF

### PR DESCRIPTION
**Pull request by kenneth-hmcts**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/CIV-10054

### Change description ###
Set Up More Information Required HWF

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by ElifDenizEsea — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/civil/enums/HwFMoreInfoRequiredDocuments.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package uk.gov.hmcts.reform.civil.enums;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+@Getter
+@RequiredArgsConstructor
+public enum HwFMoreInfoRequiredDocuments {
```
Comment (manish14836) [inline src/main/java/uk/gov/hmcts/reform/civil/enums/HwFMoreInfoRequiredDocuments.java:8]: Suggestion:- can be sorted for more readability.

File: src/main/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandler.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+@RequiredArgsConstructor
+public class MoreInformationHwfCallbackHandler extends CallbackHandler {
+
+    private static final String ERROR_MESSAGE_DOCUMENT_DATE_MUST_BE_AFTER_TODAY = "Documents date must be future date";
+
+    @Override
+    protected Map<String, Callback> callbacks() {
+        return new ImmutableMap.Builder<String, Callback>()
+            .put(callbackKey(ABOUT_TO_START), this::emptyCallbackResponse)
+            .put(callbackKey(MID, "more-information-hwf"), this::validationMoreInformation)
```
Comment (manish14836) [inline src/main/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandler.java:40]: Move out of function and create a class variable and return class variable in callback method

File: src/main/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandler.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        return Collections.singletonList(
+            CaseEvent.MORE_INFORMATION_HWF
+        );
+    }
+
+    private CallbackResponse validationMoreInformation(CallbackParams callbackParams) {
+        CaseData caseData = callbackParams.getCaseData();
+        List<String> errors = new ArrayList<>();
+        HelpWithFeesMoreInformation moreInformationData =
+            FeeType.HEARING.equals(Optional.ofNullable(caseData.getHwfFeeType()).orElse(null))
```
Comment (manish14836) [inline src/main/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandler.java:57]: Use == while comparing enums.  
FeeType.HEARING == caseData.getHwfFeeType()

File: src/main/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandler.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                ? caseData.getHelpWithFeesMoreInformationHearing()
+                : caseData.getHelpWithFeesMoreInformationClaimIssue();
+        LocalDate hwFMoreInfoDocumentDate = moreInformationData.getHwFMoreInfoDocumentDate();
+        System.out.println(hwFMoreInfoDocumentDate);
```
Comment (manish14836) [inline src/main/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandler.java:61]: Remove this System.out.println(hwFMoreInfoDocumentDate);

File: src/test/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import uk.gov.hmcts.reform.civil.service.FeatureToggleService;
+
+import java.time.LocalDate;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static uk.gov.hmcts.reform.civil.callback.CallbackType.MID;
+import static uk.gov.hmcts.reform.civil.callback.CallbackType.ABOUT_TO_SUBMIT;
+
+@ExtendWith(SpringExtension.class)
+@SpringBootTest(classes = {
```
Comment (manish14836) [inline src/test/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandlerTest.java:28]: This test doesn't need Spring testcontext. 
Please check https://github.com/hmcts/civil-service/pull/4023

File: src/test/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+class MoreInformationHwfCallbackHandlerTest extends BaseCallbackHandlerTest {
+
+    @Autowired
+    private MoreInformationHwfCallbackHandler handler;
+
+    @MockBean
+    private FeatureToggleService featureToggleService;
+
+    @Nested
+    class AboutToStartCallback {
```
Comment (manish14836) [inline src/test/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandlerTest.java:41]: Not sure if class name is correct or not as your are writing test for mid event inside this. plz check

File: src/test/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            AboutToStartOrSubmitCallbackResponse response = (AboutToStartOrSubmitCallbackResponse) handler.handle(params);
+
+            Assertions.assertThat(response.getErrors()).containsExactly("Documents date must be future date");
+        }
+
+        @Test
+        void shouldValidationMoreInformationHearing_withInvalidDate() {
+            CaseData caseData = CaseDataBuilder.builder()
+                .atStatePendingClaimIssued()
+                .build()
```
Comment (manish14836) [inline src/test/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandlerTest.java:66]: plz check if you need this build and builder, build 1 time in the end only

**Review by manish14836 — CHANGES_REQUESTED** [review]

**Review by manish14836 — DISMISSED** [review]

File: src/test/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            CallbackParams params = callbackParamsOf(caseData, MID, "more-information-hwf");
+
+            AboutToStartOrSubmitCallbackResponse response = (AboutToStartOrSubmitCallbackResponse) handler.handle(params);
+
+            Assertions.assertThat(response.getErrors()).containsExactly("Documents date must be future date");
+        }
+
+        @Test
+        void shouldValidationMoreInformationHearing_withInvalidDate() {
+            CaseData caseData = CaseData.builder()
```
Comment (portizdev) [inline src/test/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandlerTest.java:54]: Use Given, When, Then pattern pls

Comment (kenneth-hmcts) [reply src/test/java/uk/gov/hmcts/reform/civil/handler/callback/user/MoreInformationHwfCallbackHandlerTest.java:55]: Update from review

**Review by portizdev — DISMISSED** [review]

**Review by manish14836 — DISMISSED** [review]

**Review by manish14836 — DISMISSED** [review]

**Pull request merged**